internal/pkg/rerere/rerere.go

```diff
@@ -0,0 +1,224 @@
+	Options           *RetestingOptions    `json:"options,omitempty"`
+	CurrentRetryTimes int                  `json:"current_retry_times,omitempty"`
+	Time              time.Time            `json:"time,omitempty"`
+}
+
+// RetestingOptions holds options for retesting.
+type RetestingOptions struct {
+	RetestingBranch string               `json:"retesting-branch,omitempty"`
+	Retry           int                  `json:"retry,omitempty"`
+	Contexts        prowflagutil.Strings `json:"require_contexts,omitempty"`
+	Timeout         time.Duration        `json:"timeout,omitempty"`
+}
+
+// AddFlags injects retesting options into the given FlagSet.
+func (o *RetestingOptions) AddFlags(fs *flag.FlagSet) {
+	fs.StringVar(&o.RetestingBranch, "retesting-branch", defaultRetestingBranch, "Retesting target branch.")
+	fs.IntVar(&o.Retry, "retry", defaultRetestingTimes, "Retry testing times.")
+	fs.Var(&o.Contexts, "require-contexts", "Required requireContexts that must be green to merge.")
+	fs.DurationVar(&o.Timeout, "timeout", defaultTimeOut, "Test timeout time.")
+}
+
+// Validate validates retry times and contexts.
+func (o *RetestingOptions) Validate(bool) error {
+	if o.Retry <= 0 {
+		return errors.New("--retry must more than zero")
+	}
+	contexts := o.Contexts.Strings()
+	if len(contexts) == 0 {
+		return errors.New("--requireContexts must contain at least one context")
+	}
+	return nil
+}
+
+type githubClient interface {
+	ListStatuses(org, repo, ref string) ([]github.Status, error)
+	GetSingleCommit(org, repo, SHA string) (github.RepositoryCommit, error)
+	ListCheckRuns(org, repo, ref string) (*github.CheckRunList, error)
+}
+
+type gitRepoClient interface {
+	CheckoutNewBranch(branch string) error
+	Commit(title, body string) error
+	PushToCentral(branch string, force bool) error
+}
+
+// Retesting drives the current code to the test branch and keeps checking the test results.
+func Retesting(log *logrus.Entry, ghc githubClient, client gitRepoClient,
+	options *RetestingOptions, org string, repo string, spec *downwardapi.JobSpec) error {
+	log.Infof("String resting on %s/%s/branches/%s.", org, repo, options.RetestingBranch)
+	for i := 0; i < options.Retry; i++ {
+		// First time retesting we need to checkout the retesting branch.
+		if i == 0 {
+			err := client.CheckoutNewBranch(options.RetestingBranch)
+			if err != nil {
+				return err
+			}
+		}
+
+		// Commit the retry log file.
+		retestingLog := RetestingLog{
+			Job:               spec,
+			Options:           options,
+			CurrentRetryTimes: i + 1,
+			Time:              time.Now(),
+		}
+		rawLog, err := json.Marshal(retestingLog)
+		if err != nil {
+			return err
+		}
+		err = ioutil.WriteFile(defaultRetestingLogFileName, rawLog, 0600)
+		if err != nil {
+			return err
+		}
+
+		contexts := options.Contexts.Strings()
+		log.Infof("Retesting %v.", contexts)
+		err = client.Commit(fmt.Sprintf("Retesting %v", contexts), string(rawLog))
```

Comment:
Output the context name to title may cause text length to be too long.

Reply:
What do you suggest to fill in?